python/ts_scriptqueue/script_info.py

```diff
@@ -0,0 +1,254 @@
+
+    @property
+    def running(self):
+        return self._run_started and not self.done
+
+    @property
+    def done(self):
+        return self.process_task is not None and self.process_task.done()
+
+    @property
+    def index(self):
+        """Get the script's SAL index"""
+        return self._index
+
+    def run(self):
+        """Start the script running.
+
+        Raises
+        ------
+        RuntimeError
+            If the script cannot be run. Reasons could include:
+            - The script has not yet been configured.
+            - `run` was already called.
+            - The script process is done.
+        """
+        if not self.runnable:
+            raise RuntimeError("Script is not runnable")
+        self._run_started = True
+        asyncio.ensure_future(self.remote.cmd_run.start(self.remote.cmd_run.DataType()))
+
+    @property
+    def runnable(self):
+        """Can the script be run?
+
+        For a script to be runnable it must be configured,
+        the process must not have finished,
+        and ``run`` must not have been called.
+        """
+        return self.configured and not (self.done or self._run_started)
+
+    async def start_loading(self, fullpath):
+        """Start the script process and start a task that will configure
+        the script when it is ready.
+
+        fullpath : `str`, `bytes` or `os.PathLike`
+            Full path to the script.
+        """
+        initialpath = os.environ["PATH"]
+        scriptdir, scriptname = os.path.split(fullpath)
+        try:
+            os.environ["PATH"] = scriptdir + ":" + initialpath
+            self.process = await asyncio.create_subprocess_exec(scriptname, str(self.index))
+            self.timestamp = self.salinfo.manager.getCurrentTime()
+            self.process_task = asyncio.ensure_future(self.process.wait())
+            self.process_task.add_done_callback(self._run_callback)
+            self.process_task.add_done_callback(self._cleanup)
+            self.config_task = asyncio.ensure_future(self._configure())
+            self.config_task.add_done_callback(self._run_callback)
+            self.start_task.set_result(None)
+        finally:
+            os.environ["PATH"] = initialpath
+
+    def terminate(self):
+        """Terminate the script by sending SIGTERM.
+
+        Does not wait for termination to finish; to do that use::
+
+            `await script_info.process.wait()`
+
+        Returns
+        -------
+        did_terminate : `bool`
+            True if termination was requested,
+            False if there is no process or it was already finished.
+
+        Raises
+        ------
+        RuntimeError
+            If we have no information about that script, meaning
+            it already finished or was never loaded.
+        """
+        if self.process is None:
+            return False
+        if self.process.returncode is not None:
+            return False
+
+        self.process.terminate()
```

Comment:
I think that before sending a `terminate` to the subprocess, we should give the script the chance to do some cleanup. Maybe there should be a command that we could send to the script before terminating. 

For instance, scripts to take long exposures could send a stop exposure command to the camera before terminating.

Reply:
You raise a good point. Scripts can be gently shut down by sending "stop" but the script queue should probably offer a way to send that command. How about if I replace the "remove" ScriptQueue command with "stop" and "terminate"? Users should generally try "stop" before trying "terminate".

Reply:
yes, I like `stop` and `terminate` commands.

Reply:
I have implemented `stop` and `terminate` commands. If the script is running then `stop` will give the script a chance to clean up, but if that takes too long or the script is not running then it will terminate the script. So `stop` is gentler than `terminate` but should be just as reliable.